---
title: "`r paste('racoon_clip', params$workflow_type, 'processing report')`"
author: "Melina Klostermann"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc_float: TRUE
    toc: TRUE
    number_sections: yes
    fig_caption: yes
    cache: no
params:
  config: "config"
  output_dir: "output_dir" 
  snake_dir: "snake_dir"
  workflow_type: "workflow_type"

---



```{r setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(warning=FALSE, message=FALSE, eval = TRUE, cache = F, echo = FALSE,  cache.lazy = FALSE,  dpi= 20, fig.width = 20, fig.height = 10, crop = NULL) #, fig.pos = "!H", out.extra = ""
# options(knitr.table.format = "html")

# Params for direct run
# local path on MAC
# params$snake_dir <- "/Users/melinaklostermann/Documents/projects/racoon_clip/racoon_clip/racoon_clip"
# # test eCLIP
# params$output_dir <- "/Users/melinaklostermann/Documents/projects/racoon_clip/racoon_clip/tests/report_test/test_report_eCLIP"
# params$config <- read_yaml("/Users/melinaklostermann/Documents/projects/racoon_clip/racoon_clip/tests/report_test/inputs_for_report_test/eCLIP/config_test_report_eCLIP.yaml") %>% unlist()


source(paste0(params$snake_dir, "/workflow/rules/theme_html.R"))
library(ggplot2)
theme_set(theme_html())
```

```{r libraries}
library(tidyverse)
library(yaml)
library(knitr)
# library(gridExtra)
# library(kableExtra)
# library(grid)


```

# Overview of steps performed

```{r, results='asis'}
if(params$workflow_type == "peaks"){
cat("**racoon_clip peaks** was used.")
} else {
cat("**racoon_clip crosslinks** was used.")
}


```

\newline
\newline

```{r, echo=FALSE}
include_graphics(paste0(params$snake_dir, "/workflow/rules/racoon_clip_workflow_2.0.png"))

```

\newline
\newline



# Configurations {.tabset}

Here is a summary of the settings you have chosen for this analysis.

```{r config}
#getwd()

config_default <- yaml.load_file(paste0(params$snake_dir, "/config/config.yaml")) %>%
  map_dfr(., ~.x) %>%
  t(.) %>%
  as.data.frame()
#config <- yaml.load_file(params$config) %>% map_dfr(., ~.x) %>% t(.) %>% as.data.frame(.)
config <-  params$config %>% map_dfr(~.x) %>% as.data.frame(.) %>% t(.)

#params$snake_dir

colnames(config_default) <- c("config")
colnames(config) <- c("config")

# colnames(config)
n <- which(!(rownames(config_default) %in% rownames(config)))

add <- data.frame(config = config_default[n,])
rownames(add) <- rownames(config_default)[n]


config <- rbind(config, add)
#kable(config)
# config
dir <- config["wdir",]

dir <- as.character(dir)

```

```{r what_to_run}
demult <- as.logical(config["demultiplex",])
trim <- as.logical(config["adapter_trimming",])
qual_bc <- as.logical(config["quality_filter_barcodes",]) 
dedup <- as.logical(config["deduplicate",]) 
no_trim <- !any(demult, trim, qual_bc)
only_trim <- ((!demult) && trim)
qual_bc_no_demult <- ((!demult) && qual_bc )
mir <- config["experiment_type",] == "miReCLIP"

# Check if we're running peaks workflow
peaks <- (params$workflow_type == "peaks")

# Check if FastQ Screen is enabled
fastq_screen_enabled <- FALSE
if ("fastqScreen" %in% rownames(config)) {
  fastq_screen_enabled <- as.logical(config["fastqScreen",])
  if (is.na(fastq_screen_enabled)) {
    fastq_screen_enabled <- (config["fastqScreen",] == "True" || config["fastqScreen",] == "true")
  }
}

if(config["experiment_type",] == "noBarcode_noUMI"){
  dedup <- FALSE}


```

## Input files
```{r}

# df <- data.frame(parameter = c("configfile", "infiles", "barcodes_fasta", "gtf", "genome_fasta"), file = c(params$config, config[c("infiles", "barcodes_fasta", "gtf", "genome_fasta"),]))
df <- config[c(
  "infiles",
  "barcodes_fasta", 
  "gtf", 
  "genome_fasta"
), ] %>% as.data.frame(.)

rownames(df) <- c("infiles", "barcodes_fasta", "gtf", "genome_fasta")
colnames(df) <- "Input"

kable(df)  
```

## Options
```{r}
t <- data.frame(config[c(
  "demultiplex",
  "adapter_trimming", 
  "trim3",
  "quality_filter_barcodes", 
  "encode", 
  "deduplicate"
), ])
rownames(t) <- c(
  "demultiplex",
  "adapter_trimming", 
  "trim3",
  "quality_filter_barcodes", 
  "encode", 
  "deduplicate"
)
colnames(t) <- "Option"
kable(t) 
```

## Barcode settings
```{r}
t <- data.frame(config[c("barcodeLength", "minBaseQuality", "umi1_len", "umi2_len","exp_barcode_len", "samples", "encode", "encode_umi_length"),])
rownames(t) <- c("barcodeLength", "minBaseQuality", "umi1_len", "umi2_len","exp_barcode_len", "samples", "encode", "encode_umi_length")
colnames(t) <- "Setting"
kable(t) 
```

## Read filter
```{r}
t <- data.frame(config[c("min_read_length"),])
rownames(t) <- ("min_read_length")
colnames(t) <- "Setting"
kable(t) 
```

## Alignment setting (STAR)
```{r}
t <- data.frame(config[c("read_length", "outFilterMismatchNoverReadLmax", "outFilterMismatchNmax", "outFilterMultimapNmax", "outReadsUnmapped",  "outSJfilterReads", "moreSTARParameters"),])
rownames(t) <- c("read_length", "outFilterMismatchNoverReadLmax", "outFilterMismatchNmax", "outFilterMultimapNmax", "outReadsUnmapped",  "outSJfilterReads", "moreSTARParameters")
colnames(t) <- "Setting"
kable(t) 

```

## Experimental groups
```{r}
gf <- file.exists(as.character(config["experiment_group_file",]))

```

```{r eval=gf}
groups <- read.table(as.character(config["experiment_group_file",]))
colnames(groups) <- c("Group", "Sample")

kable(groups) 

```

```{r eval = !gf, results = 'asis'}
cat(paste("No groups specified. We will combine all samples for you."))
```


# {-}



# {-}



# Quality control of raw files 

Initial quality control of the raw sequencing file is performed with FastQC. Here you can see the sequencing quality and content per base.

```{r raw_quality, eval = demult}
# get name of file
name_raw_fasta <- strsplit(as.character(config["infiles",]), "/") %>% unlist(.)
l <- length(name_raw_fasta)
name_raw_fasta <- name_raw_fasta[[l]] %>% unlist(.)
name_raw_fasta <- gsub(".fastq.gz","", as.character(name_raw_fasta), fixed = T)  %>% unlist(.) %>% .[1]

# overview 
fastqc_report_raw <- read_delim(paste0(dir,"/results/fastqc/raw/", name_raw_fasta, "_fastqc/fastqc_data.txt"), delim="\t", skip = 2, n_max = 10)
kable(fastqc_report_raw)  

# images
include_graphics(paste0(dir,"/results/fastqc/raw/", name_raw_fasta, "_fastqc/Images/per_base_quality.png"))
include_graphics(paste0(dir,"/results/fastqc/raw/", name_raw_fasta, "_fastqc/Images/per_base_sequence_content.png"))


```


```{r raw_quality_demult, eval=!demult, dpi=100, include=FALSE, out.width="100%"}

fastqc_report_raw <- read_delim(paste0(dir,"/results/fastqc/raw/multiqc_data/multiqc_fastqc.txt"), delim="\t")
fastqc_report_t <- fastqc_report_raw[,c(1,5:8,11)] %>% 
  mutate(avg_sequence_length = round(avg_sequence_length, digits = 1)) %>%
  arrange(as.character(Sample))
         #Sample = substr(Sample,20,nchar(Sample)))

kable(fastqc_report_t) 



pbq <- list.files(path = paste0(dir,"/results/fastqc/raw"), pattern = "\\per_base_quality.png$", recursive = TRUE, full.names = T) 
p <- nchar(paste0(dir,"/results/fastqc/raw/"))+1
n <- nchar("/Images/per_base_quality.png")
titles_pbq <- lapply(pbq, function(x) substr(x, p, nchar(x)-n-7 ))
# titles_pbq

pbsc <- list.files(path = paste0(dir,"/results/fastqc/raw"), pattern = "\\per_base_sequence_content.png$", recursive = TRUE, full.names = T)

# list.files(path = paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming"), pattern = "\\sequence_length", recursive = TRUE, full.names = T)

# pbq <- pbq %>% map(~png::readPNG(.x, native = F) %>% rasterGrob )
# pbsc <- pbsc %>% map(~png::readPNG(.x, native = F) %>% rasterGrob )
# titles_pbq <- titles_pbq %>% map(~ggpubr::text_grob(.x))
# 
# plots <- rbind(unlist(titles_pbq), unlist(pbq), unlist(pbsc)) %>% as.list()
# 
# marrangeGrob(grobs = plots, nrow = 1, ncol =3)  

```

## Quality scores across all bases {.tabset}


```{r eval=!demult, out.width="100%", dpi = 100, results = "asis"}

for(i in 1:length(pbq)){
  cat("\n###", titles_pbq[[i]], "{-}\n")
  cat( paste0("![](", pbq[[i]], ") \n "))
}

```

## {-}

## Sequence content over all bases {.tabset}

```{r eval=!demult, out.width="100%", dpi = 100, results = "asis"}

for(i in 1:length(pbq)){
  cat("\n### ", titles_pbq[[i]], "{-}\n")
  cat( paste0("![](", pbsc[[i]], ") \n"))
}

```

## {-}

```{r eval = fastq_screen_enabled, results = 'asis'}
cat("# FastQ Screen contamination check \n")
cat("FastQ Screen was used to check for potential contamination from different organisms. The plot shows the percentage of reads mapping to each reference genome.\n")
```


```{r eval = fastq_screen_enabled && file.exists(paste0(dir,"/results/fastqscreen_multi/fastqscreen_multiqc_data/multiqc_fastq_screen.txt"))}
# Read the FastQ Screen multiqc data
fastqscreen_file <- paste0(dir,"/results/fastqscreen_multi/fastqscreen_multiqc_data/multiqc_fastq_screen.txt")

# Read header line separately to extract column names
header_line <- readLines(fastqscreen_file, n = 1)
col_names_raw <- strsplit(header_line, "\t")[[1]]

# Read the data without column names
fastqscreen_data <- read_delim(fastqscreen_file, delim="\t", col_names = FALSE, skip = 1, show_col_types = FALSE)

# Identify genome-specific columns (pairs of "counts" and "percentage")
# Structure: Sample, total_reads, [genome counts, genome percentage]*, No hits counts, No hits percentage
genome_cols <- col_names_raw[!col_names_raw %in% c("Sample", "total_reads", "No hits counts", "No hits percentage")]

# For each genome, we have two columns: "genome counts" and "genome percentage"
# Extract unique genome names by removing " counts" and " percentage" suffixes
genome_names <- unique(gsub(" (counts|percentage)$", "", genome_cols))

# Create clean column names by combining genome names with metric types
clean_col_names <- c("Sample", "total_reads")
for (genome in genome_names) {
  clean_col_names <- c(clean_col_names, paste0(genome, "_counts"), paste0(genome, "_percentage"))
}
# Add "No hits" columns at the end
clean_col_names <- c(clean_col_names, "NoHits_counts", "NoHits_percentage")

# Assign clean column names to the data
colnames(fastqscreen_data) <- clean_col_names

# Clean up sample names (remove _screen suffix if present)
fastqscreen_data$Sample <- gsub("_screen$", "", fastqscreen_data$Sample)

# Create table with original-style names for display
fastqscreen_report <- fastqscreen_data
colnames(fastqscreen_report) <- col_names_raw
fastqscreen_report$Sample <- gsub("_screen$", "", fastqscreen_report$Sample)

# Identify percentage columns for plotting (those ending with "_percentage")
percentage_cols <- grep("_percentage$", names(fastqscreen_data), value = TRUE)
genome_names_plot <- gsub("_percentage$", "", percentage_cols)

# Create a long format data frame for plotting (tidyr and stringr already loaded via tidyverse)
# Extract percentage data for each genome
plot_data <- fastqscreen_data %>%
  select(Sample, all_of(percentage_cols)) %>%
  pivot_longer(cols = all_of(percentage_cols), 
               names_to = "Genome", 
               values_to = "Percentage") %>%
  mutate(Genome = gsub("_percentage$", "", Genome))

```

```{r eval = fastq_screen_enabled && file.exists(paste0(dir,"/results/fastqscreen_multi/fastqscreen_multiqc_data/multiqc_fastq_screen.txt"))}
# Display summary table
kable(fastqscreen_report, caption = "FastQ Screen Results")
```


```{r eval = fastq_screen_enabled && file.exists(paste0(dir,"/results/fastqscreen_multi/fastqscreen_multiqc_data/multiqc_fastq_screen.txt")), fig.width=10, fig.height=6}
# Create stacked bar plot
library(ggplot2)

ggplot(plot_data, aes(x = Sample, y = Percentage, fill = Genome)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
        legend.position = "right",
        plot.title = element_text(hjust = 0.5, face = "bold")) +
  labs(title = "FastQ Screen - Genome Mapping Percentages",
       x = "Sample",
       y = "Mapping Percentage (%)",
       fill = "Genome") +
  scale_x_discrete(labels = function(x) stringr::str_wrap(x, width = 15)) +
  scale_fill_viridis_d()
```



```{r eval=qual_bc, results = 'asis'}
cat(" \n")
cat(paste("# Quality filtering in barcode region \n Sequences are filtered for a **minimum base quality of", config["minBaseQuality",], "** within the barcode region to ensure correct sample and duplicate assignment.\n"))
cat(" \n")
```

```{r eval=qual_bc, results = 'asis'}
cat("## Read numbers before and after the quality filtering step \n ")
```


```{r}
bc_filter_one_file <- (demult && qual_bc)
bc_filter_multi_file <- ((!demult) && qual_bc)

```


```{r barcode_filter1, eval=demult}

fastqc_report_filtered <- read_delim(paste0(dir, "/results/fastqc/filtered/filtered_fastqc/fastqc_data.txt"),
  delim = "\t", 
  skip = 2, 
  n_max = 10
)
read_numbers <- data.frame( steps = c("raw", "after_barcode_quality_filter"),
                            `number of reads` = c(fastqc_report_raw$Value[4], fastqc_report_filtered$Value[4]))

kable(read_numbers)  

```

```{r barcode_filter1_multi, eval=qual_bc_no_demult }

fastqc_report_filtered <- read_delim(paste0(dir,"/results/fastqc/filtered/multiqc_data/multiqc_fastqc.txt"), delim="\t")

read_numbers <- data.frame( Sample = fastqc_report_raw$Sample,
                            "Raw reads" = fastqc_report_raw$`Total Sequences`, 
                            "Filtered reads" = fastqc_report_filtered$`Total Sequences`) %>% 
  arrange(as.character(Sample))

kable(read_numbers)  

```


```{r eval=qual_bc, results = 'asis'}
cat("## Sequencing quality and content per base after the quality filtering step \n")
```

```{r barcode_filter2, eval=bc_filter_one_file, results = 'asis'}
cat("### Quality scores across all bases \n")
include_graphics(paste0(dir,"/results/fastqc/filtered/filtered_fastqc/Images/per_base_quality.png"))
cat(" \n")
cat(" \n")
```


```{r barcode_filter2.2, eval=bc_filter_one_file, results = 'asis'}
cat(" \n")
cat("\n ### Sequence content over all bases \n")
cat(" \n")
include_graphics(paste0(dir,"/results/fastqc/filtered/filtered_fastqc/Images/per_base_sequence_content.png"))
cat(" \n")

```

```{r eval=qual_bc_no_demult , results = 'asis'}
cat("### Quality scores across all bases {.tabset} \n")
```


```{r barcode_filter2_multi, eval=qual_bc_no_demult , out.width="100%", dpi = 100, results = "asis"}

fastqc_report_filtered_t <- fastqc_report_filtered[,c(1,5:7,11)] 

# %>% 
#   mutate(avg_sequence_length = round(avg_sequence_length, digits = 1)) %>% 
#   arrange(as.character(Sample))


kable(fastqc_report_filtered_t)


# include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_per_base_sequence_quality_plot_1.png"))
# include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_per_sequence_gc_content_plot_Percentages.png"))
# 
# include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_sequence_length_distribution_plot_1.png"))



pbq <- list.files(path = paste0(dir,"/results/fastqc/filtered"), pattern = "\\per_base_quality.png$", recursive = TRUE, full.names = T) 
p <- nchar(paste0(dir,"/results/fastqc/filtered/")) + 1
n <- nchar("/Images/per_base_quality.png")
titles_pbq <- lapply(pbq, function(x) substr(x, p ,nchar(x)-n-7 ))
# titles_pbq

pbsc <- list.files(path = paste0(dir,"/results/fastqc/filtered"), pattern = "\\per_base_sequence_content.png$", recursive = TRUE, full.names = T)

# list.files(path = paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming"), pattern = "\\sequence_length", recursive = TRUE, full.names = T)

# pbq <- pbq %>% map(~png::readPNG(.x, native = F) %>% rasterGrob )
# pbsc <- pbsc %>% map(~png::readPNG(.x, native = F) %>% rasterGrob )
# titles_pbq <- titles_pbq %>% map(~ggpubr::text_grob(.x))
# 
# plots <- rbind(unlist(titles_pbq), unlist(pbq), unlist(pbsc)) %>% as.list()
# 
# marrangeGrob(grobs = plots, nrow = 1, ncol =3)  


# pmap(list(a = titles_pbq, b = pbq, c = pbsc), function(a,b,c) print(marrangeGrob(grobs = list(b,c), title = a, nrow=2)   ))

```





```{r test, out.width="100%", dpi = 100, results = 'asis', eval=qual_bc_no_demult }
for(i in 1:length(pbq)){
  cat("####", titles_pbq[[i]], "{-}\n")
  cat( paste0("![](", pbq[[i]], ") \n"))
  cat("\n")
}
cat("### {-} \n")

```

```{r eval=qual_bc_no_demult , results = 'asis'}
cat("\n")
cat("### Sequence content over all bases {.tabset}")
cat("\n")
```

```{r out.width="100%", dpi = 100, results = "asis", eval=qual_bc_no_demult }
for(i in 1:length(pbsc)){
  cat("####", titles_pbq[[i]], "{-}\n")
  cat( paste0("![](", pbsc[[i]], ") \n"))
  cat("\n")
}

```

```{r eval=qual_bc_no_demult , results = 'asis'}
cat("### {-}")
```

```{r}
# demult <- (config["demuliplex",] == "TRUE")
# trim <- (config["adapter_trimming",] == "TRUE")
# qual_bc <- (config["quality_filter_barcodes",] == "TRUE")
```



```{r eval=demult, results = 'asis'}
cat("# Demultiplexing and quality control of individual samples \n
Samples are demultiplexed with Flexbar and adapters are trimmed in the same step.")

```


```{r demultiplexed_samples, out.width="100%", dpi = 100, results = "asis", eval=demult}
fastqc_report_filtered <- read_delim(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_data/multiqc_fastqc.txt"), delim="\t")
fastqc_report_filtered_t <- fastqc_report_filtered[,c(1,5:7,11)] %>% 
  #mutate(avg_sequence_length = round(avg_sequence_length, digits = 1)) %>% 
  arrange(as.character(Sample))
         #Sample = substr(Sample,20,nchar(Sample)))

kable(fastqc_report_filtered_t) 


# include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_per_base_sequence_quality_plot_1.png"))
# include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_per_sequence_gc_content_plot_Percentages.png"))
# 
# include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_sequence_length_distribution_plot_1.png"))


# pbq <- list.files(path = paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming"), pattern = "\\per_base_quality.png$", recursive = TRUE, full.names = T) 
# p <- nchar(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/"))+1
# n <- nchar("/Images/per_base_quality.png")
# titles_pbq <- lapply(pbq, function(x) substr(x, p, nchar(x)-n-7 ))
# # titles_pbq
# 
# pbsc <- list.files(path = paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming"), pattern = "\\per_base_sequence_content.png$", recursive = TRUE, full.names = T)
# 
# # list.files(path = paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming"), pattern = "\\sequence_length", recursive = TRUE, full.names = T)
# 
# # pbq <- pbq %>% map(~png::readPNG(.x, native = F) %>% rasterGrob )
# # pbsc <- pbsc %>% map(~png::readPNG(.x, native = F) %>% rasterGrob )
# # titles_pbq <- titles_pbq %>% map(~ggpubr::text_grob(.x))
# # 
# # plots <- rbind(unlist(titles_pbq), unlist(pbq), unlist(pbsc)) %>% as.list()
# # 
# # marrangeGrob(grobs = plots, nrow = 1, ncol =3)  
# 
# for(i in 1:length(pbq)){
#   cat(paste0("\n  **", titles_pbq[[i]], "** \n"))
#   cat( paste0("![](", pbq[[i]], ") \n"))
#   cat( paste0("![](", pbsc[[i]], ") \n"))
# }

# pmap(list(a = titles_pbq, b = pbq, c = pbsc), function(a,b,c) print(marrangeGrob(grobs = list(b,c), title = a, nrow=2)   ))

  
```

```{r eval=F, results = 'asis'}
# asis_output("# Adapter trimming \\n")
print("## Mutations in barcodes /n Mutations can occur during PCR amplification or read sequencing. Mutations within the gene sequence are dealt with during the star alignment step. However, mutations in the barcode sequence can cause the read to be assigned to the wrong sample. As a result, a read with a single or multiple mutated barcode will not be assigned to any sample. Here is an overview of the distance between the detected barcodes and the specified sample barcodes.")
```

```{r barcode_distances, out.width="100%", eval=FALSE}
barcodes_found <- read_table(paste0(dir,"/results/barcode_filter/barcodes_detected.txt"), col_names = c("n", "barcode"))


barcodes_fasta <- read_table(as.character(config["barcodes_fasta",]), col_names = F)
barcodes_fasta <- data.frame(sample = barcodes_fasta[c(1:nrow(barcodes_fasta) %% 2 != 0),],
                             barcode = barcodes_fasta[c(1:nrow(barcodes_fasta) %% 2 == 0),])
colnames(barcodes_fasta) <- c("sample", "barcode")


barcodes_fasta <- barcodes_fasta %>% mutate(., exp_barcode = substr(barcode, as.numeric(config["umi1_len",])+1, as.numeric(config["umi1_len",])+as.numeric(config["exp_barcode_len",]) ) )
kable(barcodes_fasta, caption = "sample barcodes") 

barcodes_found <- barcodes_found %>%
  rowwise %>%
  mutate( mismatch_0 = vcountPattern(barcode, barcodes_fasta$barcode, max.mismatch=0) %>% sum(),
          mismatch_1 = vcountPattern(barcode, barcodes_fasta$barcode, max.mismatch=1) %>% sum(),
          mismatch_2 = vcountPattern(barcode, barcodes_fasta$barcode, max.mismatch=2) %>% sum(),
          mismatch_3 = vcountPattern(barcode, barcodes_fasta$barcode, max.mismatch=3) %>% sum(),
          mismatch_4 = vcountPattern(barcode, barcodes_fasta$barcode, max.mismatch=4) %>% sum(),
          mismatch_5 = vcountPattern(barcode, barcodes_fasta$barcode, max.mismatch=5) %>% sum(),
          dist = case_when(mismatch_0 > 0 ~ 0,
                           mismatch_1 > 0 ~ 1,
                           mismatch_2 > 0 ~ 2,
                           mismatch_3 > 0 ~ 3,
                           mismatch_4 > 0 ~ 4,
                           mismatch_5 > 0 ~ 5,
                           ),
          # dist = factor(dist, levels = dist),
          barcode = factor(barcode, levels = barcode),
          dist = as.character(dist))  %>%
  arrange(dist)

head(barcodes_found)

ggplot(barcodes_found[c(1:100),], aes(x=barcode, y = n, fill = dist))+
  geom_col()+
  theme(text=element_text(size=50))+
  scale_fill_viridis_d()


ggplot(barcodes_found[ barcodes_found$dist != 0,], aes(x=barcode, y = log(n), fill = as.factor(dist)))+
  geom_col()+
  theme(text=element_text(size=50))+
  scale_fill_viridis_d()





```

# Adapter trimming

```{r eval=only_trim, results = 'asis'}
cat(paste("\n  The adapters are trimmed with Flexbar which would be used for demultiplexing on multiplexed data. This should ensure compatibility between multiplexed and demultiplexed data sets. \\n 
   The adapters used are specified with the adapter_file parameter. The default file contains standard Illumina and eCLIP adapters. The adapters are trimmed from the 3' end of the read (--adapter-trim-end RIGHT), with an allowed error of 0.1 (--adapter-error-rate 0.1). The adapter must overlap the read by at least 1nt (--adapter-min-overlap 1). Adapter trimming can be done in multiple cycles as specified by the user, here", config["adapter_cycles",], "(adapterCycles, see configurations). Reads shorter than", config["min_read_length",],"are discarded (min_read_length, see configurations). \\n
   UMIs are written into the read header with (--umi-tags). If the data is from ENCODE, UMIs are already removed from the read and written to the beginning of the header. As umi-tools dedup, which is used here for deduplication, expects the UMI to be at the end of the read, encode: True can be specified to move the UMI from the beginning of the read header to the end (see configurations).
        \\n"))

```

## Adapter content before trimming

Position of adapters found in the reads. If adapters are found early in the reads, this indicates over digestion of the samples.

```{r, results='asis'}

if (qual_bc & !demult){
 f = paste0(dir,"/results/fastqc/filtered/multiqc_plots/png/mqc_fastqc_adapter_content_plot_1.png")
}else{
  if(!demult){
    f= paste0(dir,"/results/fastqc/raw/multiqc_plots/png/mqc_fastqc_adapter_content_plot_1.png")
    }else{
    f=paste0(dir,"/results/fastqc/raw/filtered_fastqc/Images/adapter_content.png")
 }}
    
    
if(file.exists(f)){
    include_graphics(f)
  }else{
    cat("\n \n  **No adapter content was detected in the raw files. **") }



```



```{r eval=only_trim, results = 'asis'}
cat("## Adapter content after trimming \n")

if(file.exists(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_adapter_content_plot_1.png"))){
  include_graphics(paste0(dir,"/results/fastqc/samples_after_demultiplexing_or_trimming/multiqc_plots/png/mqc_fastqc_adapter_content_plot_1.png"))
}else{
  cat("\n \n  **Trimming successful, no adapter contents left. **")
}

```

```{r eval = no_trim, results = 'asis'}
cat("No trimming has been performed.")
```

# Alignment

The genome annotation is indexed using STAR --runMode genomeGenerate. The merged file of chimeric reads is then aligned to the genome annotation using STAR --runMode alignReads. For STAR settings, see Configurations. The resulting bam files are indexed using samtools index.

```{r eval=!mir }
star_out <- list.files(path = paste0(dir,"/results/aligned"), pattern = "*Log.final.out$", recursive = TRUE, full.names = T) %>% map(~read_delim(.x, delim="\t", col_names = c("type", "number")))

star_mir_out_names <-  list.files(paste0(dir,"/results/aligned"), pattern = "*Log.final.out$", recursive = TRUE) %>%
  gsub("chimeric_", "", . ) %>%
  gsub(".Log.final.out", "",  . )


names(star_out) <- star_mir_out_names 

r <- c(5,8,9,23,24,25,26,28:33)
star_out2 <- star_out %>% map(., ~.x[r,]) %>%
  map_dfr( ~.x$number) 


names <- star_out[[1]]$type[c(5,8,9,23,24,25,26,28:33)]
star_out2 <- star_out2 %>% as.data.frame(.)
rownames(star_out2) <- substr(names,1, nchar(star_out[[1]]$type)-2)

# star_out2 <- star_out2 %>%
#   arrange(as.character(Sample)) %>%
#   .[, r]

kable(star_out2)
```


```{r eval =mir}

star_out <- list.files(path = paste0(dir, "/results/aligned"), pattern = "*Log.final.out$", recursive = TRUE, full.names = T) %>% map(~ read_delim(.x, delim = "\t", col_names = c("type", "number")))

star_mir_out_names <-  list.files(paste0(dir,"/results/aligned"), pattern = "*Log.final.out$", recursive = TRUE) %>%
  gsub("chimeric_", "", . ) %>%
  gsub(".Log.final.out", "",  . )

names(star_out) <- star_mir_out_names 

r <- c(5,8,9,23,24,25,26,28:33)
star_out2 <- star_out %>% map(., ~.x[r,]) %>%
  map_dfr( ~.x$number) 


names <- star_out[[1]]$type[c(5,8,9,23,24,25,26,28:33)]
rownames(star_out2) <- substr(names,1, nchar(star_out[[1]]$type)-2)

# star_out2 <- star_out2 %>%
#   arrange(as.character(Sample)) %>%
#   .[, r]

kable(as.data.frame(star_out2) )
```

```{r, dpi=100 }
gg.df <- t(star_out2) %>% as.data.frame(.) %>% .[,c(1,2,6) ]

gg.df$Sample <- rownames(gg.df)
colnames(gg.df) <- c("reads",  "uniquely_mapped_reads", "mapped_reads", "Sample")


gg.df <- gg.df %>% mutate( reads = as.numeric(reads),
                           uniquely_mapped_reads = as.numeric(uniquely_mapped_reads),
                           mapped_reads = as.numeric(mapped_reads) + uniquely_mapped_reads
                           )

gg.df <- gg.df %>% reshape2::melt()

gg.df$variable <- factor(gg.df$variable, levels = c("reads", "mapped_reads", "uniquely_mapped_reads"))



ggplot(gg.df, aes( x = Sample, y = as.numeric(value), fill = variable))+
  geom_col(position = "dodge", stat = "idenity")+
  theme(#axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1),
        legend.title= element_blank(),
        legend.position = "top")+
  ylab("Number of reads")+
  coord_flip()+
  scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 25)) +
  scale_fill_viridis_d()



```


```{r eval=dedup, results = 'asis'}
# asis_output("# Adapter trimming \\n")
cat("# Deduplication \n
Reads are deduplicated with umi_tools dedup --extract-umi-method read_id --method unique.")
```

# Obtaining crosslinks

The (deduplicated) bam files are then converted to bed files using bedtools bamtobed. The reads are shifted 1nt upstream (5' direction) with bedtools shift -m 1 -p -1 because the UV crosslink should be positioned 1nt earlier. 
\n
To allow visualisation in a genome browser, the 1nt crosslink bed file is also converted to a .bigWig file using bedGraphToBigWig. These bigWig files are then merged with bigWigMerge by the experiment groups specified by the user.

```{r dpi = 100, eval=dedup}

umitools_out <- list.files(path = paste0(dir,"/results/aligned"), pattern = "*out.duprm.log$", recursive = TRUE, full.names = T) %>%
  map(~read_delim(.x, delim="\t", skip = 57, col_names = c("invalue")))

# umitools_out <- map(umitools_out, ~ umitools_out[(length(umitools_out) - 6):length(umitools_out), ])

umitools_out <- map(umitools_out, ~mutate(.x, measure = substr(invalue, 29, nchar(invalue)) %>% strsplit(., split = ":")) )


umitools_out = map(umitools_out, ~mutate(.x,
 measure=modify_if(measure,
                           .p = is_empty,
                           .f = ~NA_character_))
)

umitools_out <- map(umitools_out, ~ rowwise(.x) %>%
  mutate(
    type = measure[[1]],
    number = measure[[length(measure)]]
  ) %>%
  .[(nrow(.) - 5):(nrow(.) - 2), ])

umitools_out_names <-  list.files(path = paste0(dir,"/results/aligned"), pattern = "*out.duprm.log$", recursive = TRUE) %>%
  gsub("chimeric_", "", . ) %>%
  gsub(".Aligned.sortedByCoord.out.duprm.log", "", . )

names(umitools_out) <- umitools_out

umitools_out_2 <- map_dfr(umitools_out, ~.x$number)
umitools_out_2 <- t(umitools_out_2) %>% as.data.frame()

colnames(umitools_out_2) <- c("Reads", "Crosslink events (without PCR duplicates)", "Crosslinked nucleotides", "Mean crosslink events per crosslinked nucleotide")
umitools_out_2$Sample <- umitools_out_names
rownames(umitools_out_2) <- umitools_out_names

kable(umitools_out_2[,1:4])

gg.umitools_out_2 <- umitools_out_2 %>% reshape2::melt(id.vars = c("Mean crosslink events per crosslinked nucleotide", "Sample"))
gg.umitools_out_2$value <- as.numeric(gg.umitools_out_2$value)

ggplot(gg.umitools_out_2, aes(x = Sample, y = value, fill = variable))+
  geom_col(position = "dodge")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        legend.title= element_blank(),
        legend.position = "top")+
  ylab("Number of reads")+
  coord_flip()+
  scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 25)) +
  scale_fill_viridis_d()

```

```{r eval=peaks, results='asis'}
cat("# Peak calling \n")
cat("Peaks are called with pureclip on the merged bam files of the sample groups (or all samples if no groups are specified). \n\n")
```

```{r eval=peaks, results='asis'}
# Function to count lines in peak bed files
count_peak_lines <- function(peak_dir) {
  # Find all .bed files in the peaks directory
  peak_files <- list.files(path = peak_dir, pattern = "\\.bed$", full.names = TRUE, recursive = TRUE)
  
  if (length(peak_files) == 0) {
    return(data.frame(File = "No peak files found", Peaks = 0, stringsAsFactors = FALSE))
  }
  
  # Count lines in each file using system wc -l command
  peak_counts <- sapply(peak_files, function(file) {
    if (file.exists(file)) {
      # Use system command wc -l for efficient line counting
      wc_result <- system(paste("wc -l", shQuote(file)), intern = TRUE)
      # Extract the number from the wc output (first element before the filename)
      line_count <- as.numeric(strsplit(trimws(wc_result), "\\s+")[[1]][1])
      return(line_count)
    } else {
      return(0)
    }
  })
  
  # Create a data frame with file names and counts
  peak_summary <- data.frame(
    File = basename(names(peak_counts)),
    Peaks = as.numeric(peak_counts),
    stringsAsFactors = FALSE
  )
  
  # Sort by file name
  peak_summary <- peak_summary[order(peak_summary$File), ]
  rownames(peak_summary) <- NULL
  
  return(peak_summary)
}

# Count peaks if directory exists
peaks_dir <- paste0(dir, "/results/peaks")
if (dir.exists(peaks_dir)) {
  peak_summary <- count_peak_lines(peaks_dir)
  

  cat("Number of peaks identified in each sample/group: \n\n")}
```

```{r eval=peaks, results='asis'}
if (dir.exists(peaks_dir)) {
  print(kable(as.data.frame(peak_summary)))
  
  # Also create a simple bar plot if there are peaks
  if (nrow(peak_summary) > 0 && any(peak_summary$Peaks > 0)) {
    cat("\n\n")
    
    ggplot(peak_summary, aes(x = File, y = Peaks)) +
      geom_col(fill = "steelblue", alpha = 0.7) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      labs(x = "Sample/Group", y = "Number of Peaks", title = "Peak Counts by Sample/Group") +
      coord_flip()
  }
} else {
  cat("Peak directory not found. Peaks may not have been called yet or the directory structure may be different. \n\n")
}
```

# Session info

```{r}
sessionInfo()

```



